### DISCUSSION TITLE
Mock auth() instance Jest

### DISCUSSION THREAD

**Discussion by bad-encoder** [invertase/react-native-firebase#5851] [Q&A]
Hi, when I try to mock provider methods - everything works , but it doesn't work with methods of auth() instance
How do I do this so I don't get an error:
` TypeError: (0 , _auth.default) is not a function`

test: 
```
jest.mock('@react-native-firebase/auth', () => ({
     GoogleAuthProvider: {
      credential: jest.fn().mockReturnValue('123'),
    },
    // how to mock auth() instance or signInWithGoogle method? 
}));
```
code:
```
import auth from '@react-native-firebase/auth';
    import {GoogleSignin} from '@react-native-community/google-signin';
    
    export const GoogleLogin = async function (): Promise<void | string> {
      // Get the users ID token
      const {idToken} = await GoogleSignin.signIn();
    
      // Create a Google credential with the token
      const googleCredential = auth.GoogleAuthProvider.credential(idToken);
    
      try {
        auth().signInWithCredential(googleCredential);
      } catch (e) {
        console.log(e);
      }
    };
```
```

**Comment by mikehardy** [comment]
Hi there! We use jest as part of our testing strategy internally, and our jest.setup.ts may serve as an example
Have you seen it?

https://github.com/invertase/react-native-firebase/blob/main/jest.setup.ts

Hopefully that gets you going

**Comment by bad-encoder** [comment]
Thank you, Mike, for reply. There is a little confusion, it seemed necessary to simply mock the result of the function, but something is not easy for me to implement... I don't need to mock the function if I set up the jest setup this way? Maybe you have an example of a test written with the signInWithCredential function?

**Reply by mikehardy** [reply]
You will most likely still need to do specific mocks for things you want to test, but that setup file will get you past the part where `@react-native-firebase/app` (our core module) keeps trying to auto-load the native modules, and it demonstrates *how*/*where* to put mocks easily for the methods you want to test

for instance this line right here https://github.com/invertase/react-native-firebase/blob/d0d9bd84c1ea231437e9129c62ed68ccd3a4a181/jest.setup.ts#L44 handles mocking `addIdTokenListener`, if you were to add a new entry there for `signInWithCredential` and have it do whatever you need, your mock method would be the one called when it you called `firebase.auth().signInWithCredential(...)` in your code anywhere

I believe with this information you should be at the "try it and see..." stage, so, try it and see :-)

**Reply by bad-encoder** [reply]
Perfect, i've got it, now it is works
Thank you
But how about the case, when I wanna mock function with dynamic result in specific cases(promise rejects e.g)? How can I mock the function inside test-case? 
Can I do it here in some way? 
```
jest.mock('@react-native-firebase/auth', () => ({
     GoogleAuthProvider: {
      credential: jest.fn().mockReturnValue('123'),
    },
    // how to mock auth() instance or signInWithGoogle method? 
}));
```

**Reply by mikehardy** [reply]
"try it and see" :sweat_smile: :smile: - I don't do this so have never tried. In the most brutal way it appears you could save the mocked module placed into the native modules array from jest setup and replace the whole thing with a temporary one then put it back, but what you propose may work and be more elegant. With jest literally anything is possible, so, whatever works and seems right for your project

**Comment by tomskopek** [comment]
@bad-encoder I realize this is quite late but for anyone else coming through this way, maybe this will help.

Personally I found the [jest.setup referenced](https://github.com/invertase/react-native-firebase/blob/main/jest.setup.ts) to be quite confusing, it's not how I'm used to doing mocks so I didn't really know how to make use of it. That said, testing and mocking is not my forte!

I did something like this (we're using ts-jest):

```js
// Component.test.tsx
import { mocked } from 'ts-jest/utils';

import auth from '../../../utils/firebaseAuth'; // In my case I'm importing from here, importing the library directly should also work
jest.mock('../../../utils/firebaseAuth');
const mockedAuth = mocked(auth, true);

beforeEach(() => {
  mockedAuth.mockReset();
})

it('should do something', () => {
  mockedAuth.mockReturnValue({
    currentUser: null, // here you can put what you need
    onAuthStateChanged: jest.fn(),
  } as unknown as FirebaseAuthTypes.Module);
  
  // ... rest of your test
})
```
